### PR TITLE
Remote environments: New & Updated Components

### PR DESCRIPTION
**Pull request by wbreza**
Adds the following new components to support remote environments:

## What's New
- **environment.Manager**
  - The listing, retrieval and saving of environments utilizing `environment.DataStore` contructs
- **environment.LocalFileDataStore**
  - The existing local file implementation of environment state
- **environment.StorageBlobDataStore**
  - Default implementation of storing environment state in azure blob storage
- **config.FileConfigManager**
  - Updated to consume file paths for config storage
  
## Updated
- ConfigManager updated to use `io.Reader` and `io.Writer` respectively

> [!NOTE]  
> This does not update azd functionality to consume new environment components

### CODE REVIEW

File: eng/pipelines/release-cli.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@ resources:
     - repository: azure-sdk-build-tools
       type: git
       name: internal/azure-sdk-build-tools
-      ref: refs/tags/azure-sdk-build-tools_20230828.2
+      ref: refs/tags/azure-sdk-build-tools_20230829.1
```
Comment (danieljurek) [inline eng/pipelines/release-cli.yml:6]: This should take on whatever value is in `main`. 

Provided your local `main` branch is up to date this can be done by running: 

```bash 
git checkout main -- eng/pipelines/release-cli.yml
```

File: eng/pipelines/release-vscode.yml [hunk 2]
```diff
@@ -3,7 +3,7 @@ resources:
     - repository: azure-sdk-build-tools
       type: git
       name: internal/azure-sdk-build-tools
-      ref: refs/tags/azure-sdk-build-tools_20230828.2
+      ref: refs/tags/azure-sdk-build-tools_20230829.1
```
Comment (danieljurek) [inline eng/pipelines/release-vscode.yml:6]: This should take on whatever value is in `main`. 

Provided your local `main` branch is up to date this can be done by running: 

```bash 
git checkout main -- eng/pipelines/release-vscode.yml
```

File: eng/pipelines/release-cli.yml (hunk 1 above)
Comment (wbreza) [reply eng/pipelines/release-cli.yml]: I didn't make this change - likely a rebase/timing issue.

File: cli/azd/pkg/environment/manager.go [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+)
+
+// Manager is the interface used for managing instances of environments
+type Manager interface {
+	Create(ctx context.Context, spec Spec) (*Environment, error)
+	LoadOrCreateInteractive(ctx context.Context, name string) (*Environment, error)
+	List(ctx context.Context) ([]*Description, error)
+	Get(ctx context.Context, name string) (*Environment, error)
+	Save(ctx context.Context, env *Environment) error
+	Reload(ctx context.Context, env *Environment) error
```
Comment (vhvb1989) [inline cli/azd/pkg/environment/manager.go:48]: IMO, we should get rid of this method.

I think that any component currently calling `Reload()` should be doing a `load + delta-comparison` instead or reloading.

IIRC, we use this method for cases where the environment can be changed from some external source, like a hook.
When a component `A` invokes a hook, it needs to call `Reload()` after the hook returns to get any changes to the env. But the component could also load the env again and compare it against the one within the component and then take decisions based on that.

That's probably orthogonal to this PR, but I'd start the note/thread here, in case folks want to talk about it.

**Review by vhvb1989 — COMMENTED** [review]
a few comments.

File: cli/azd/pkg/azsdk/storage/storage_blob_client.go [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+	return nil
+}
+
+// createClient creates a new blob client and caches it for future use
+func (bc *blobClient) createClient(ctx context.Context) (*azblob.Client, error) {
+	if bc.client != nil {
+		return bc.client, nil
+	}
+
+	credential, err := bc.authManager.CredentialForCurrentUser(ctx, nil)
```
[the remark this answers is not in the document]
Comment (wbreza) [reply cli/azd/pkg/azsdk/storage/storage_blob_client.go:214]: @vhvb1989 is there some limitation with `CredentialForCurrentUser` that it would only restrict to tenant owners?  I am currently under the impression that it would use the permissions of the logged in user when reading/writing to blob storage vs requiring a different authentication strategy of manual connection of SAS tokens.

File: cli/azd/pkg/azsdk/storage/storage_blob_client.go [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+	blobOptions := &azblob.ClientOptions{
+		ClientOptions: *coreOptions,
+	}
+
+	if bc.config.Endpoint == "" {
+		bc.config.Endpoint = DefaultBlobEndpoint
+	}
+
+	serviceUrl := fmt.Sprintf("https://%s.%s", bc.config.AccountName, bc.config.Endpoint)
+	client, err := azblob.NewClient(serviceUrl, credential, blobOptions)
```
[the remark this answers is not in the document]
Comment (wbreza) [reply cli/azd/pkg/azsdk/storage/storage_blob_client.go:232]: Should be possible - i will look into it.

File: cli/azd/pkg/environment/manager.go (hunk 3 above)
Comment (wbreza) [reply cli/azd/pkg/environment/manager.go:48]: Good call out but I'm going to keep this for now and we can follow up with a separate PR if we intent to change the reloading behavior.

File: cli/azd/pkg/config/file_config_manager.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+// FileConfigManager provides the ability to load, parse and save azd configuration files
+type FileConfigManager interface {
+	// Saves the azd configuration to the specified file path
+	// Path is automatically created if it does not exist
+	Save(config Config, filePath string) error
+
+	// Loads azd configuration from the specified file path
+	Load(filePath string) (Config, error)
+}
```
[the remark this answers is not in the document]
Comment (wbreza) [reply cli/azd/pkg/config/file_config_manager.go:19]: Yes, i modified the root `config.Manager` to use `io.Reader` and `io.Writer` respectively and then created this `config.FileConfigManager` for use cases where data is persisted to files.

File: cli/azd/.vscode/cspell-azd-dictionary.txt [hunk 7]
```diff
@@ -113,6 +116,7 @@ otlp
 otlpconfig
 otlptrace
 otlptracehttp
+paketobuildpacks
```
[the remark this answers is not in the document]
Comment (wbreza) [reply cli/azd/.vscode/cspell-azd-dictionary.txt]: This was from before the rebase.

File: cli/azd/pkg/azsdk/storage/storage_blob_client.go [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+	// Upload uploads a blob to the configured storage account container.
+	Upload(ctx context.Context, blobPath string, reader io.Reader) error
+
+	// Delete deletes a blob from the configured storage account container.
+	Delete(ctx context.Context, blobPath string) error
+
+	// Items returns a list of blobs in the configured storage account container.
+	Items(ctx context.Context) ([]*Blob, error)
+}
```
[the remark this answers is not in the document]
Comment (wbreza) [reply cli/azd/pkg/azsdk/storage/storage_blob_client.go:46]: I prefer a simpler API for only exposing the bits we need to integrate with blob storage vs taking a dependency directly on the blob client.

File: cli/azd/pkg/azsdk/storage/storage_blob_client.go [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+		for _, container := range page.ContainerItems {
+			if *container.Name == bc.config.ContainerName {
+				exists = true
+				break
+			}
+		}
+	}
+
+	if !exists {
```
Comment (ellismg) [inline cli/azd/pkg/azsdk/storage/storage_blob_client.go:166]: Some of the other language SDKs provide a CreateIfNotExists method which works by trying to create the container and then ignoring the ContainerAlreadyExists error that comes back: 

https://github.com/Azure/azure-sdk-for-net/blob/d4e55222514a6c310db25fd2df0415cd0aacf708/sdk/storage/Azure.Storage.Blobs/src/BlobContainerClient.cs#L910-L1119

Perhaps that approach is better here?  The above does have a TOCTOU bug, since the container could be created (or I guess removed) between when we start the page operation and when we complete it.

Might be reasonable to just follow the SDK Team's lead here.

File: cli/azd/pkg/azsdk/storage/storage_blob_client.go [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+		}
+	}
+
+	return nil
+}
+
+// createClient creates a new blob client and caches it for future use
+func NewBlobSdkClient(
+	ctx context.Context,
+	authManager *auth.Manager,
```
Comment (ellismg) [inline cli/azd/pkg/azsdk/storage/storage_blob_client.go:179]: Could we just take a TokenCredential here and have the caller supply it or does that not work great for our DI strategy?  It seems preferable to me to take the more specific abstraction if we can.

File: cli/azd/pkg/environment/data_store.go [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+type RemoteKind string
+
+const (
+	RemoteKindAzureStorage RemoteKind = "AzureStorage"
+)
+
+type DataStore interface {
+	// Gets the path to the environment .env file
+	Path(env *Environment) string
```
Comment (ellismg) [inline cli/azd/pkg/environment/data_store.go:19]: Nit: Consider adding a prefix like `Env` so this is `EnvPath`.  It would provide symmetry with `ConfigPath` and make it a little clearer, I think, about what actually is being returned.

File: cli/azd/pkg/azsdk/storage/storage_blob_client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ellismg) [reply cli/azd/pkg/azsdk/storage/storage_blob_client.go]: The nice thing about using managed identity here is that we don't have to worry about storing any secrets in our configuration.  I think in practice that folks are not going to use Anonymous access here (especially since you'd have to allow anonymous writes, which is super scary).  Perhaps supporting SAS as well would be reasonable, but then we'd have this problem of securing the SAS token itself, which we don't have a great story for.

The downside is that Azure RBAC is a little tricky, and so you might find that you need to grant RBAC permissions to the principal that is doing the deployment.  In an ideal world, it might be nice for these to be separatable, but I feel like starting with just a "you must use managed identity and Azure RBAC" is fine.  If it ends up being such an issue that we have to figure out how to go add support for SAS, I think there's a reasonable path forward for us to do this.

**Review by ellismg — APPROVED** [review]
I think you struck a good balance here on refactoring/new features.  There's still some duplicated code we might want to address longer term (for example, I think we have now duplicated the code that adjusts the .env file so numbers are not quoted across our two backends) but landing this with the current factoring seems great to me.

A few small comments - I feel most strongly about taking TokenCredential instead of AuthManager in `NewSdkBlobClient`.

Great work here, @wbreza!  Excited to see the abstraction and the blob storage implementation.

File: cli/azd/pkg/azsdk/storage/storage_blob_client.go (hunk 10 above)
Comment (wbreza) [reply cli/azd/pkg/azsdk/storage/storage_blob_client.go:179]: If we go down the route of using TokenCredential here and have the client injected via IoC then I'm ok with that as long as we can use the authManager to create the token credential by default.

**Review by vhvb1989 — APPROVED** [review]

File: cli/azd/pkg/azsdk/storage/storage_blob_client.go (hunk 9 above)
Comment (wbreza) [reply cli/azd/pkg/azsdk/storage/storage_blob_client.go:166]: The Go sdk doesn't include a `CreateIfNotExists` error.  I tried just ignoring the case where the container already exists but this slows things down because it forces the internal retry strategy to retry multiple times causing a more significant delay.

Going to keep as is for the moment.

File: cli/azd/pkg/environment/manager.go [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+
+// Get returns the environment instance for the specified environment name
+func (m *manager) Get(ctx context.Context, name string) (*Environment, error) {
+	localEnv, err := m.local.Get(ctx, name)
+	if err != nil {
+		if m.remote == nil {
+			return nil, err
+		}
+
+		remoteEnv, err := m.remote.Get(ctx, name)
```
Comment (weikanglim) [inline cli/azd/pkg/environment/manager.go:280]: What if the remote copy is behind the local copy? This ends up overriding the local copy implicitly.

File: cli/azd/pkg/environment/data_store.go [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+// DataStore is the interface for the interacting with the persistent storage of environments.
+
+type RemoteKind string
+
+const (
+	RemoteKindAzureStorage RemoteKind = "AzureStorage"
+)
+
+type DataStore interface {
+	// Gets the path to the environment .env file
```
Comment (weikanglim) [inline cli/azd/pkg/environment/data_store.go:18]: Should this intertace be exposed publicly? If so, we should maybe consider explaining what a `path` is.

File: cli/azd/pkg/state/config.go [hunk 14]
```diff
@@ -0,0 +1,12 @@
+package state
+
+// Config is the state configuration for an azd project
+type Config struct {
```
Comment (weikanglim) [inline cli/azd/pkg/state/config.go:4]: Not a huge fan of splaying the `azd` configuration schema across multiple packages. I would prefer if the configuration lives in one root package for the config file, and we can certainly create new structs for domain-specific handling if needed.

**Comment by weikanglim** [comment]
There are some interesting implications raised by the design that may be good to think about. It's likely that this may have been discussed privately, I just wanted to make sure we keep track of the decisions somewhere for posterity.

1. What is `azd`'s stance on how remote environments should be used? Are CI environments expected to point to remote states?  Is `azd provision` now an inherently stateful operation? It used to be possible for this to be done locally, we've expanded it to enable this posibility in production. Before going down this path, we should really ensure that everyone is onboard with the expctations. There should be official guidance offered by `azd` here.
2. What sort of persistence strategy are we offering with remote environments? I'm not quite understanding the decision in `manager.go`. It is unclear to me what the proposed strategy is when reading the code. It seems like we expect dual-read and dual-write in most places. Dual-write and dual-read is one of the hardest implementations to get right, and I wonder if we need to be a little bit more detailed about how we expect the system to work as a whole. This avoids non-trivial bugs, and also allows for long-term maintainability.

On the general coding side:

1. I'm not a huge fan of how we're continually shifting into app-heavy components that are coupled in ways that are not immediately clear. It used to be that `environment` is a standalone package that can be used to manipulate environments. This change makes it impossible without dragging in other concepts like `azd/ioc`.
2. There is over-abstraction / introducing of new types here that I'm a little concerned about. What is a `FileConfigManager`, and why is it reshared across `environment`? What real-world behavior does `FileConfigManager` offer that warrants this? Our codebase is becoming increasing complex, and in my view, for not all the right reasons. Personally, this is scary because if we ever become slightly divergent in implementation (due to requirements) on how `environment` and "user config", this may not be easily caught. This could be a naming this. The question is what is the general logic that we should share? For me, the arbitrary key-value handling of a config file is trivial, and I could see us not sharing the implementation. On the flipside, if there are some unique features, it should be called out in a documented type that clearly represents the functionality. `flexibleConfig`, `jsonConfig`, something of the sorts that indicates its real-world usefulness which is the spirit of go code.

To summarize on my proposal of the way forward:

1. Let's document azd's official stance on the design decisions raised.
2. It may be too late on the coding side, and we can move forward as needed if impeding progress, but I think we as a team need to somewhat align on a coding style.

File: eng/pipelines/release-cli.yml [hunk 15]
```diff
@@ -156,6 +156,8 @@ stages:
               AZD_TEST_TENANT_ID: $(arm-tenant-id)
               AZD_TEST_AZURE_SUBSCRIPTION_ID: $(SubscriptionId)
               AZD_TEST_AZURE_LOCATION: eastus2
+              AZD_TEST_REMOTE_STATE_ACCOUNT: saazdremotestate
+              AZD_TEST_REMOTE_STATE_CONTAINER: remote-env-test
```
Comment (danieljurek) [inline eng/pipelines/release-cli.yml:160]: We've avoided static resources in favor of standing up resources per-run and tearing them down. 

Since the storage account lives outside the scope of the resources created for the test run we should create a separate resource group and storage account within that group and then tear those down. We do have a [script](https://github.com/Azure/azure-sdk-tools/blob/main/eng/common/TestResources/New-TestResources.ps1) which does that for the Azure SDK Engineering System, but I'm not sure how coupled it is to our `sdk/<service>` directory layout and other engsys features (like environment variable export).

File: cli/azd/pkg/environment/data_store.go (hunk 13 above)
Comment (wbreza) [reply cli/azd/pkg/environment/data_store.go:18]: This has been renamed to `EnvPath` to be more clear.

File: cli/azd/pkg/environment/manager.go (hunk 12 above)
Comment (wbreza) [reply cli/azd/pkg/environment/manager.go:280]: The goal here is similar to terraform remote state.  The state is always synced for every state operation in azd.  Any time an environment save is performed the change is persisted locally and then immediately saved to the remote as well.  @ellismg and I spoke about this privately but agree we should document this behavior.

File: eng/pipelines/release-cli.yml (hunk 15 above)
Comment (wbreza) [reply eng/pipelines/release-cli.yml:160]: I agree on this strategy.  I will do a fast follow PR to this to figure out how we want to handle these types of scenarios.

**Comment by wbreza** [comment]
> To summarize on my proposal of the way forward:
> 
> 1. Let's document azd's official stance on the design decisions raised.
> 2. It may be too late on the coding side, and we can move forward as needed if impeding progress, but I think we as a team need to somewhat align on a coding style.

Thanks for the feedback @weikanglim. These can be good topics in our next engineering discussion.
